Add GeneratedStaticResourceProcessor

mcruzdev opened the pull request:
Fixes #39735 


cc @FroMage @cescoffier @ppalaga

ia3andy commented on extensions/vertx-http/deployment/src/main/java/io/quarkus/vertx/http/deployment/StaticResourcesProcessor.java:
```diff
@@ -82,6 +86,31 @@ public void nativeImageResource(Optional<StaticResourcesBuildItem> staticResourc
+            List<GeneratedStaticResourceBuildItem> generatedStaticResourceBuildItems,
+            BuildProducer<GeneratedResourceBuildItem> generatedResourceBuildItemBuildProducer,
+            BuildProducer<NativeImageResourceBuildItem> nativeImageResourcesProducer,
+            BuildProducer<AdditionalStaticResourceBuildItem> additionalStaticResourceBuildItemBuildProducer,
```
Only for prod mode, we have two options:
1. we could use `additionalStaticResourceBuildItemBuildProducer`
2. we could create a new StaticHandler

I think `additionalStaticResourceBuildItemBuildProducer` seems better.

ia3andy commented:
In https://github.com/quarkusio/quarkus/blob/main/core/deployment/src/main/java/io/quarkus/deployment/builditem/GeneratedResourceBuildItem.java#L9-L11

We should change the comment to recommend using `GeneratedStaticResourceBuildItem` which takes care of this.

ia3andy commented:
In `AdditionalStaticResourceBuildItem` we should also explain that it's not made for generated resources and to use `GeneratedStaticResourceBuildItem` instead.

`AdditionalStaticResourceBuildItem` should only be for prod, and maybe we should even remove the processing for dev (to avoid any possible conflict):
https://github.com/quarkusio/quarkus/blob/main/extensions/vertx-http/deployment/src/main/java/io/quarkus/vertx/http/deployment/StaticResourcesProcessor.java#L49-L51

FroMage commented:
In any case, this is going to need tests.

ia3andy commented on extensions/vertx-http/deployment/src/main/java/io/quarkus/vertx/http/deployment/GeneratedStaticResourcesProcessor.java:
```diff
@@ -0,0 +1,90 @@
+            generatedResourceBuildItem.produce(
+                    new GeneratedResourceBuildItem(resourceOnGeneratedResources,
+                            staticResource.getContent(), false));
+            additionalStaticResourcesProducer.produce(
+                    new AdditionalStaticResourceBuildItem(staticResource.getPublicPath(), false));
+            if (!devOrTest) {
+                additionalStaticResourcesProducer.produce(
+                        new AdditionalStaticResourceBuildItem(staticResource.getPublicPath(), false));
```
It seems you are adding it twice here

ia3andy commented on extensions/vertx-http/deployment-spi/src/main/java/io/quarkus/vertx/http/deployment/spi/GeneratedStaticResourceBuildItem.java:
```diff
@@ -0,0 +1,28 @@
+/**
+ * GeneratedStaticResourceBuildItem aims to be used by extensions to generated static resources into
+ * <code>META-INF/resource</code> directory. This means to be used just for {@link io.quarkus.runtime.LaunchMode#TEST} or
+ * {@link io.quarkus.runtime.LaunchMode#DEVELOPMENT} mode.
```
It is not only for dev or test, it is also for production.

Could you add a bit of explanation on what it does behind the scene: It takes care of creating what needed for routing in dev, test, and prod mode and native (you don't create any other BuildItem for this to be working).

ia3andy commented on extensions/vertx-http/deployment-spi/src/main/java/io/quarkus/vertx/http/deployment/spi/GeneratedStaticResourceBuildItem.java:
```diff
@@ -0,0 +1,28 @@
+import io.quarkus.builder.item.MultiBuildItem;
+
+/**
+ * GeneratedStaticResourceBuildItem aims to be used by extensions to generated static resources into
+ * <code>META-INF/resource</code> directory. This means to be used just for {@link io.quarkus.runtime.LaunchMode#TEST} or
```
```suggestion
 * GeneratedStaticResourceBuildItem aims to be used by extensions to generate static resources. Those resources will be served on the given publicPath. It is NOT necessary to create the file on disk. This means to be used just for {@link io.quarkus.runtime.LaunchMode#TEST} or
```

ia3andy commented on extensions/vertx-http/deployment/src/main/java/io/quarkus/vertx/http/deployment/GeneratedStaticResourcesProcessor.java:
```diff
@@ -0,0 +1,90 @@
+    private static final int ROUTE_ORDER = RouteConstants.ROUTE_ORDER_BEFORE_DEFAULT + 60;
+
+    @BuildStep
+    @Record(ExecutionTime.RUNTIME_INIT)
```
I think you should separate this in two buildsteps, one for creating the different buildItems and another for the runtime init that uses the recorder. You can choose to create the Map in a new internal BuildItem or create it in runtime init

ia3andy commented on extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/handlers/DevClasspathStaticHandler.java:
```diff
@@ -0,0 +1,137 @@
+        }
+    }
+
+    private String getClasspathResource(String name) {
```
I don't think you should use a String here, could you keep it as a byte array or byte buffer depending on what vertx wants?

ia3andy reviewed (changes requested):
Overall I think you are on the right way 👍

Just added a few comments here and there...

ia3andy commented on extensions/vertx-http/deployment/src/main/java/io/quarkus/vertx/http/deployment/GeneratedStaticResourcesProcessor.java:
```diff
@@ -0,0 +1,103 @@
+                generatedStaticResourceBuildItem.getPublicPath());
+    }
+
+    /**
+     * Give a {@link GeneratedStaticResourceBuildItem}, write its content on the {@code generatedResource} path.
+     *
+     * @param generatedResource {@link GeneratedStaticResourceBuildItem} instance.
+     * @param buildDir {@link Path} representing where the {@link GeneratedStaticResourceBuildItem} will be written.
+     */
+    private static void writeGeneratedResourceOnBuildDir(GeneratedStaticResourceBuildItem generatedResource,
+            Path buildDir) {
+        final Path targetPath = buildDir.resolve(generatedResource.getPublicPath());
+        try {
+            Files.deleteIfExists(targetPath);
+            Files.createDirectories(targetPath.getParent());
+            Files.write(targetPath, generatedResource.getContent());
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
```
```suggestion
```

Not used anymore right?

mcruzdev replied:
Yes, I agree with you

ia3andy commented on extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/handlers/DevClasspathStaticHandler.java:
```diff
@@ -0,0 +1,138 @@
+    private void doHandle(RoutingContext context) {
+        String resolvedPath = resolvePath(context);
+        String path = resolvedPath.endsWith("/") ? resolvedPath.concat(this.indexPage) : resolvedPath;
+
```
We could add some debug logging here

ia3andy commented on extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/handlers/DevClasspathStaticHandler.java:
```diff
@@ -0,0 +1,138 @@
+        if (this.generatedResources.contains(removeStartSlashes(path))) {
+            compressIfNeeded(context, path);
+            byte[] content = getClasspathResourceContent(path);
+            if (content != null) {
```
I guess here should be a warning, if the path is found but their is not content no?

ia3andy reviewed (changes requested):
You are really close I think

ia3andy commented on extensions/vertx-http/deployment-spi/src/main/java/io/quarkus/vertx/http/deployment/spi/GeneratedStaticResourcesBuildItem.java:
```diff
@@ -0,0 +1,23 @@
+/**
+ * This build items aims to be a collection of all {@link GeneratedStaticResourceBuildItem} produced by extensions.
+ */
+public final class GeneratedStaticResourcesBuildItem extends SimpleBuildItem {
```
I think this should be moved to deployment as it's not meant to be used externally.

ia3andy commented on extensions/vertx-http/deployment/src/main/java/io/quarkus/vertx/http/deployment/GeneratedStaticResourcesProcessor.java:
```diff
@@ -0,0 +1,103 @@
+        }
+
+        if (!generatedResourcesPaths.isEmpty()) {
+            generatedStaticResourcesMap.produce(new GeneratedStaticResourcesBuildItem(generatedResourcesPaths));
```
```suggestion
            generatedStaticResources.produce(new GeneratedStaticResourcesBuildItem(generatedResourcesPaths));
```

ia3andy commented:
Could we add a DevModeTest similar to this:
https://github.com/quarkiverse/quarkus-web-bundler/blob/main/deployment/src/test/java/io/quarkiverse/web/bundler/test/WebBundlerDevModeTest.java

gsmet commented on extensions/vertx-http/deployment-spi/src/main/java/io/quarkus/vertx/http/deployment/spi/AdditionalStaticResourceBuildItem.java:
```diff
@@ -3,11 +3,16 @@
 
 /**
- * Meant to be used by extensions that generate resource into {@code META-INF/resources}.
+ * Meant to be used by extensions that generate resource into {@code META-INF/resources} on
+ * {@link io.quarkus.runtime.LaunchMode#NORMAL} mode, if you want to generate resource into {@code META-INF/resources}
+ * use {@link GeneratedStaticResourceBuildItem} instead.
```
I'm finding this a bit confusing. I read this sentence a few times and I don't understand exactly when I should use one and not the other?

We need to be sure people will use the right one so we need things to be extremely clear. BTW, maybe @cescoffier can chime in to improve that as IIUC the general idea came from him.

ia3andy replied:
Yeah it's not clear, `AdditionalStaticResourceBuildItem` should not be used by external extensions anymore, we keep it here for backward compat. We could duplicate it in deployment and mark this one as @Deprecated, but that would be too complex no?

gsmet commented on extensions/vertx-http/deployment-spi/src/main/java/io/quarkus/vertx/http/deployment/spi/GeneratedStaticResourceBuildItem.java:
```diff
@@ -0,0 +1,31 @@
+ */
+public final class GeneratedStaticResourceBuildItem extends MultiBuildItem {
+
+    private final String publicPath;
```
I think we should probably stick to `path`. I don't think we have this notion of `public` elsewhere (but please correct me if I'm wrong).
I see it has been part of the initial proposal but I think it's a bit incorrect as I suppose it might not be entirely public if security is added on top of it.

I would use `path` and copy the paragraph we have above:
```
 * The value of {@code path} should be prefixed with {@code '/'} and is assumed to be a path
 * under {@code 'META-INF/resources'}.
```

gsmet commented on extensions/vertx-http/deployment/src/main/java/io/quarkus/vertx/http/deployment/GeneratedStaticResourcesProcessor.java:
```diff
@@ -0,0 +1,80 @@
+package io.quarkus.vertx.http.deployment;
+
+import java.util.*;
```
Please do not use star imports.

mcruzdev replied:
Thank you for looking it :)

gsmet commented on extensions/vertx-http/deployment/src/main/java/io/quarkus/devui/deployment/GeneratedStaticResourcesBuildItem.java:
```diff
@@ -0,0 +1,24 @@
+/**
+ * This build items aims to be a collection of all {@link GeneratedStaticResourceBuildItem} produced by extensions.
+ */
+public final class GeneratedStaticResourcesBuildItem extends SimpleBuildItem {
```
What's the purpose of this one?
Asking because it seems to be a bit redundant with `GeneratedStaticResourceBuildItem`. Except if it aggregates things from other build items  but it doesn't look obvious from the javadoc?

I don't think we have this sort of things elsewhere so except if it has a value I missed, I would drop it.

mcruzdev replied:
> What's the purpose of this one?

This means be a collection of `GeneratedStaticresourceBuildItem`, I removed it and generated another `@BuildStep` with `@Record`.

gsmet commented on extensions/vertx-http/deployment/src/main/java/io/quarkus/vertx/http/deployment/GeneratedStaticResourcesProcessor.java:
```diff
@@ -0,0 +1,80 @@
+            BuildProducer<RouteBuildItem> routes, GeneratedStaticResourcesRecorder generatedStaticResourcesRecorder,
+            LaunchModeBuildItem launchModeBuildItem) {
+
+        if (generatedStaticResources.isEmpty())
```
We usually recommend to have braces even in this case.
It's easy to get things broken if not.

gsmet reviewed (commented):
Nice contribution!

I had a small look as I was wondering what it was doing and I added a few comments. Probably worth waiting for @cescoffier 's feedback before acting on the ones that are not obvious.

ia3andy commented on extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/handlers/DevClasspathStaticHandler.java:
```diff
@@ -0,0 +1,164 @@
+    private void doHandle(RoutingContext context) {
+
+        if (context.request().method().equals(HttpMethod.HEAD)) {
+            context.response().end();
```
I think the HEAD method should also return the content size, all headers but not the actual content.

ia3andy commented:
This is getting really close :)

mcruzdev commented:
> Nice contribution!
> 
> I had a small look as I was wondering what it was doing and I added a few comments. Probably worth waiting for @cescoffier 's feedback before acting on the ones that are not obvious.

I changed the documentation being more direct to avoid confusion, let me know WDYT?

cescoffier commented on extensions/vertx-http/deployment/src/test/resources/styles.css:
```diff
@@ -0,0 +1,3 @@
+.quarkus-blue {
```
Can we use a dummy text file instead?

mcruzdev replied:
Hi @cescoffier, I will use the file that has been used by StaticHandler test, WDYT?

cescoffier commented on extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/GeneratedStaticResourcesRecorder.java:
```diff
@@ -0,0 +1,59 @@
+    }
+
+    public static void setHotDeploymentResources(List<Path> changedPaths) {
+
```
why do we have this method?

mcruzdev replied:
Miss code, TY!

cescoffier commented on extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/handlers/DevClasspathStaticHandler.java:
```diff
@@ -0,0 +1,178 @@
+
+    private static final Logger LOG = Logger.getLogger(DevClasspathStaticHandler.class);
+    private static final Set<HttpMethod> ALLOWED_HTTP_METHODS = Set.of(HttpMethod.GET, HttpMethod.HEAD, HttpMethod.OPTIONS);
+    private static final int HTTP_STATUS_OK = 200;
```
Don't we have these constant somewhere else?

mcruzdev replied:
I made a search for  `= 200` in the project but there is no option. I have access to `package io.netty.handler.codec.http.HttpResponseStatus`, makes sense to use it?

cescoffier commented on extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/handlers/DevClasspathStaticHandler.java:
```diff
@@ -0,0 +1,178 @@
+
+    @Override
+    public void handle(RoutingContext context) {
+        if (!ALLOWED_HTTP_METHODS.contains(context.request().method())) {
```
any reason to do that? The route has been registered for these methods anyway.

ia3andy commented on extensions/vertx-http/deployment-spi/src/main/java/io/quarkus/vertx/http/deployment/spi/GeneratedStaticResourceBuildItem.java:
```diff
@@ -0,0 +1,37 @@
+        return this.path;
+    }
+
+    public byte[] getContent() {
```
It's already the case with GeneratedResourceBuildItem right?

cescoffier replied:
Yes, but we need to make sure we do not run out of memory

ia3andy replied:
@cescoffier how could we do that? We could add a size limit? But what would be a good max value?

ia3andy replied:
Generated Resource are all resources which where not usable as is from the sources, so it could include image and other things.

After thinking a bit more about it, this is just an way in the GeneratedResourceBuildItem which already exist, we should add a limit in there if we really want to, but I think we are fine since in prod there is a cache as we are not changing the flow at all.

In dev, the problem is just for loading those resources, in a follow up PR I will add a cache, which gets cleared after a successfull doScan. 

This way we have a proper solution for all cases.

FroMage replied:
One thing we might generate is a JS API for REST endpoints. It could be large. As long as they end up on disk, I don't think we need to have a size limit.

ia3andy commented on extensions/vertx-http/deployment/src/main/java/io/quarkus/vertx/http/deployment/GeneratedStaticResourcesProcessor.java:
```diff
@@ -0,0 +1,134 @@
+
+        if (launchModeBuildItem.getLaunchMode().isDevOrTest()) {
+
+            routes.produce(RouteBuildItem.builder().orderedRoute("/*", ROUTE_ORDER,
```
We could use a `BuildProducer<DefaultRouteBuildItem> defaultRoutes` instead right?

mcruzdev replied:
I think no, I tried here and I am receiving: 

```shell
Caused by: io.quarkus.builder.BuildException: Build failure: Too many default routes.
```

ia3andy replied:
Ok so you don't have any other choice than what you did

ia3andy replied:
or maybe allow 2 default routes in devmode

mcruzdev commented on extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/handlers/DevClasspathStaticHandler.java:
```diff
@@ -0,0 +1,178 @@
+            return null;
+        }
+        try {
+            try (InputStream inputStream = resource.openStream()) {
```
I need your help here @cescoffier, Any idea or example to make this one non blocking?

ia3andy replied:
@mcruzdev I guess you should have the answer in the StaticHandler

ia3andy replied:
@mcruzdev you should be able to use something like this:
https://vertx.io/docs/apidocs/io/vertx/core/http/HttpServerResponse.html#send-io.vertx.core.streams.ReadStream-

mcruzdev replied:
Thank you @ia3andy, I try it

mcruzdev commented on core/deployment/src/main/java/io/quarkus/deployment/builditem/GeneratedResourceBuildItem.java:
```diff
@@ -6,9 +6,17 @@ public final class GeneratedResourceBuildItem extends MultiBuildItem {
+     * It is currently used by Kogito to get serving of static resources in Dev Mode by Vert.x
+     * <br>
+     *
+     * @deprecated If you want to serve static resources in Dev mode use
```
Makes sense, I will remove this one

mcruzdev commented on extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/GeneratedStaticResourcesRecorder.java:
```diff
@@ -0,0 +1,59 @@
+    public static final String META_INF_RESOURCES = "META-INF/resources";
+    private final RuntimeValue<HttpConfiguration> httpConfiguration;
+    private final HttpBuildTimeConfig httpBuildTimeConfig;
+    private Set<String> compressMediaTypes = Set.of();
```
Hi @cescoffier, yes, it is expected.

mcruzdev commented on extensions/vertx-http/deployment/src/test/java/io/quarkus/vertx/http/StaticResourcesCachingDisabledTest.java:
```diff
@@ -6,18 +6,19 @@
 
     @RegisterExtension
-    final static QuarkusUnitTest test = new QuarkusUnitTest()
+    final static QuarkusProdModeTest prod = new QuarkusProdModeTest()
```
@ia3andy, I turned to `QuarkusUnitTest` again and it is working!

ia3andy commented:
@mcruzdev you also need to edit this:
https://github.com/quarkusio/quarkus/blob/2bfd9cbb63b4511a36edc981ff8886f6b32a05d5/extensions/vertx-http/deployment/src/main/java/io/quarkus/vertx/http/deployment/devmode/NotFoundProcessor.java#L58

In order to display the generated routes in the 404

ia3andy commented on extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/handlers/DevClasspathStaticHandler.java:
```diff
@@ -0,0 +1,178 @@
+    public DevClasspathStaticHandler(Set<String> generatedResources, DevClasspathStaticHandlerOptions options) {
+        this.generatedResources = generatedResources;
+        this.compressedMediaTypes = options.getCompressMediaTypes();
+        this.currentClassLoader = Thread.currentThread().getContextClassLoader();
```
This is the same code used in the current StaticHandler: https://github.com/quarkusio/quarkus/blob/main/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/StaticResourcesRecorder.java#L73

ia3andy replied:
This is only used in dev so we are alright. 
Also it's the exact same logic.

ia3andy commented on extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/handlers/DevClasspathStaticHandler.java:
```diff
@@ -0,0 +1,178 @@
+        if (cl == null) {
+            cl = getClass().getClassLoader();
+        }
+        // when running on substratevm (graal) the access to class loaders
```
Should we remove the comment? or change this method? This is taken from the vertx code.

ia3andy commented on extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/handlers/DevClasspathStaticHandler.java:
```diff
@@ -0,0 +1,176 @@
+                future.fail(e);
+            }
+        }, asyncResult -> {
+            if (asyncResult.succeeded()) {
```
@mcruzdev here, this if should be replaced by `if(asyncResult.result() != null)`  you should add another if for the succeed and in case of failure:
```java
context.fail(res.cause());
```

ia3andy commented:
This is an awesome contribution by @mcruzdev on his free time 🙏🚀

This PR looks good to me, I'll test it with the web-bundler today. @gsmet @cescoffier @FroMage please give your final review on the code part. It would be great if we could merge and increment so if there are some issue which can be fixed in other following PRs, let's merge this one and create issues.

ia3andy commented on extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/handlers/DevClasspathStaticHandler.java:
```diff
@@ -0,0 +1,177 @@
+                handleAsyncResultSucceeded(context, asyncResult, path);
+            } else {
+                context.fail(asyncResult.cause());
+            }
```
Where did we change the TCCL?

ia3andy reviewed (approved):
This is good to be merged targeting 3.12, I'll add support for Path based generated resources in another PR.

ia3andy commented:
@mcruzdev could you fix CI?

mcruzdev committed:
Add GeneratedStaticResourcesProcessor for handling generated static resources

FroMage commented on extensions/vertx-http/deployment/src/main/java/io/quarkus/vertx/http/deployment/GeneratedStaticResourcesProcessor.java:
```diff
@@ -0,0 +1,137 @@
+            generatedResourceBuildItem.produce(
+                    new GeneratedResourceBuildItem(generatedStaticResourceLocation,
+                            generatedStaticResource.getContent(), false));
+            // We can't use the vert.x StaticHandler for tests as it doesn't support 'quarkus' protocol
```
We could also fix it to support it.

ia3andy replied:
Yes, we can always resolve to it if that gets fixed without changing the GeneratedStaticResourceBuildItem api

FroMage commented on extensions/vertx-http/deployment/src/main/java/io/quarkus/vertx/http/deployment/StaticResourcesProcessor.java:
```diff
@@ -35,15 +36,19 @@ public class StaticResourcesProcessor {
         Set<StaticResourcesBuildItem.Entry> paths = getClasspathResources();
-        for (AdditionalStaticResourceBuildItem bi : additionalStaticResources) {
-            paths.add(new StaticResourcesBuildItem.Entry(bi.getPath(), bi.isDirectory()));
+        if (!launchModeBuildItem.getLaunchMode().isDevOrTest()) {
```
Probably add a comment explaining where they're added in the case of dev/test mode?

FroMage commented on extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/handlers/DevClasspathStaticHandler.java:
```diff
@@ -0,0 +1,175 @@
+
+        compressIfNeeded(context, path);
+
+        context.vertx().executeBlocking(future -> {
```
I would object that this is using blocking reads for reading what is most likely a local file, while `StaticHandlerImpl` uses the non-blocking file reading…
But this is only for DEV mode, so… Feels weird though having to resort to a worker thread just to read a file in DEV mode…

ia3andy replied:
Yes, this is true, maybe we can know if this is a local file and do a different processing in that case?

ia3andy replied:
anyway, I will make a few change in the next few days (to add `Path` support to GeneratedStaticResourceProcessor), this might be our solution.

FroMage reviewed (commented):
LGTM. Nice tests. Some remarks, but not blocking.

Perhaps the missing bit is to mention this in https://quarkus.io/guides/writing-extensions but I looked briefly where it could fit, and found no mention of generating static resources. So, perhaps not required.

gsmet commented:
Folks, I will let you decide if it's ready to get merged and merge it if so. Thanks.

mcruzdev commented:
> LGTM. Nice tests. Some remarks, but not blocking.
> 
> Perhaps the missing bit is to mention this in https://quarkus.io/guides/writing-extensions but I looked briefly where it could fit, and found no mention of generating static resources. So, perhaps not required.

Thank you @FroMage, I will solve the comments soon!

ia3andy commented:
@mcruzdev it's merged, provide PRs for the comments and stuff, I will work on re-adding AdditionalStaticResources to make sure we don't break existing dev-modes and adding Path support.

ia3andy commented:
@mcruzdev Congratulations! 🎉

mcruzdev commented:
Thank you for all support @ia3andy!
